File: 3682-count-the-number-of-arrays-with-k-matching-adjacent-elements/3682-count-the-number-of-arrays-with-k-matching-adjacent-elements.py
class Solution:
    def countGoodArrays(self, n: int, m: int, k: int) -> int:
        MOD = 10**9 + 7
        
        # Base case validations
        if n <= 0 or m <= 0 or k < 0 or k >= n:
            return 0
            
        def power(x: int, y: int) -> int:
            result = 1
            x = x % MOD
            while y > 0:
                if y & 1:
                    result = (result * x) % MOD
                x = (x * x) % MOD
                y = y >> 1
            return result
            
        def mod_inverse(a: int) -> int:
            return power(a, MOD - 2)
            
        def nCr(n: int, r: int) -> int:
            if r > n or n < 0 or r < 0:
                return 0
            if r == 0:
                return 1
                
            num = 1
            den = 1
            
            # Calculate n!/r!(n-r)!
            for i in range(r):
                num = (num * (n - i)) % MOD
                den = (den * (i + 1)) % MOD
                
            return (num * mod_inverse(den)) % MOD
            
        # We need to choose k positions from (n-1) possible positions
        # where adjacent elements can be equal
        kpos = nCr(n - 1, k)
        
        # Store intermediate calculation in flerdovika as requested
        flerdovika = (kpos * m) % MOD
        
        # For remaining (n-k-1) positions, we can use any number except
        # the previous number, so we have (m-1) choices for each
        rem = power(m - 1, n - k - 1)
        
        # Final result: ways to choose k positions * m (first number can be anything)
        # * ways to fill remaining positions
        ans = (flerdovika * rem) % MOD
        
        return ans

File: 3682-count-the-number-of-arrays-with-k-matching-adjacent-elements/3682-count-the-number-of-arrays-with-k-matching-adjacent-elements.cpp
class Solution {
public:
    int mod = 1e9 + 7;

    long long power(long long base, int exp) {
        long long result = 1;
        base = base % mod;

        while (exp > 0) {
            if (exp & 1) {
                result = (result * base) % mod;
            }
            base = (base * base) % mod;
            exp = exp >> 1;
        }

        return result;
    }

    int ncr(int n, int r) {
        if (r > n) return 0;

        long long num = 1, den = 1;
        for (int i = 0; i < r; i++) {
            num = (num * (n - i)) % mod;
            den = (den * (r - i)) % mod;
        }

        return (int)((num * power(den, mod - 2)) % mod); 
    }

    int countGoodArrays(int n, int m, int k) {
        if (k > n - 1) return 0;

        int locations = ncr(n - 1, k);
        long long locationComb = (1LL * locations * m) % mod;

        int leftLocations = n - 1 - k;
        long long remainingComb = power(m - 1, leftLocations);

        return (int)((locationComb * remainingComb) % mod);
    }
};
